scipy/spatial/src/distance_wrap.c

```diff
@@ -87,172 +108,163 @@ DEFINE_WRAP_CDIST(hamming, double)
 DEFINE_WRAP_CDIST(jaccard, double)
 DEFINE_WRAP_CDIST(sqeuclidean, double)
 
-DEFINE_WRAP_CDIST(yule_bool, char)
-
-
-static NPY_INLINE double *mahalanobis_dimbuf(Py_ssize_t n)
-{
-    double *dimbuf;
-    dimbuf = calloc(n, 2 * sizeof(double));
-    if (!dimbuf) {
-        PyErr_Format(PyExc_MemoryError, "could not allocate %zd * %zd bytes",
-                     n, 2 * sizeof(double));
-    }
-    return dimbuf;
-}
-
-
-static PyObject *cdist_mahalanobis_wrap(PyObject *self, PyObject *args, PyObject *kwargs) {
-  PyArrayObject *XA_, *XB_, *covinv_, *dm_;
+DEFINE_WRAP_CDIST(dice, char)
+DEFINE_WRAP_CDIST(hamming, char)
+DEFINE_WRAP_CDIST(jaccard, char)
+DEFINE_WRAP_CDIST(kulsinski, char)
+DEFINE_WRAP_CDIST(rogerstanimoto, char)
+DEFINE_WRAP_CDIST(russellrao, char)
+DEFINE_WRAP_CDIST(sokalmichener, char)
+DEFINE_WRAP_CDIST(sokalsneath, char)
+DEFINE_WRAP_CDIST(yule, char)
+
+static PyObject *cdist_cosine_double_wrap(PyObject *self, PyObject *args, 
+                                               PyObject *kwargs) {
+  PyArrayObject *XA_, *XB_, *dm_;
   int mA, mB, n;
-  double *dm, *dimbuf;
+  double *dm;
   const double *XA, *XB;
-  const double *covinv;
-  static char *kwlist[] = {"XA", "XB", "dm", "VI", NULL};
+  static char *kwlist[] = {"XA", "XB", "dm", NULL};
   if (!PyArg_ParseTupleAndKeywords(args, kwargs, 
-            "O!O!O!O!:cdist_mahalanobis_wrap", kwlist,
+            "O!O!O!:cdist_cosine_double_wrap", kwlist,
             &PyArray_Type, &XA_, &PyArray_Type, &XB_, 
-            &PyArray_Type, &dm_, &PyArray_Type, &covinv_)) {
+            &PyArray_Type, &dm_)) 
+  {
     return 0;
   }
   else {
     NPY_BEGIN_THREADS_DEF;
     NPY_BEGIN_THREADS;
     XA = (const double*)XA_->data;
     XB = (const double*)XB_->data;
-    covinv = (const double*)covinv_->data;
     dm = (double*)dm_->data;
     mA = XA_->dimensions[0];
     mB = XB_->dimensions[0];
     n = XA_->dimensions[1];
 
-    dimbuf = mahalanobis_dimbuf(n);
-    if (!dimbuf) {
+    double * norms_bufA = _cosine_norms_buf(mA);
```

Comment:
`_cosine_norms_buf` raises a Python exception if it cannot allocate its memory, but this call is inside a `NPY_BEGIN_THREADS` / `NPY_END_THREADS`, so its running with the GIL released. So if we ever run out of memory we will top it with a segfault...

Consider the following redesign idea:
 * Get rid of the `_cosine_norms_buf` function.
 * Change the signature of `cdist_cosine` to remove `norms_bufA` and `norms_bufB` and return an `int`.
 * Have `cdist_cosine` allocate (and release) the memory for its buffers, but rather than set the Python exception return e.g. `-1` if an allocation fails, `0` if all goes smoothly.
 * Here you would do:

````
if (cdist_cosine(...) < 0) {
    NPY_END_THREADS;
    PyErr_NoMemory();
    return NULL;
}
```